### PR TITLE
Skip behavior change for templated tests (#4426)

### PR DESCRIPTION
**Pull request by olesz**
Implementing the enhancement described in this issue: https://github.com/robotframework/robotframework/issues/4426

Executed the acceptance tests. This change does not seem to break any of them.

### CODE REVIEW

**Review by pekkaklarck — CHANGES_REQUESTED** [review]
This looks mostly good, but there's a bug related to checking syntax errors, graceful terminations and test timeouts too late, I also think the added new exceptions for reporting these special cases are pretty complicated. I don't think they are needed, at least if we don't need to include skip messages in tests final message. As I commented separately, I have a simpler prototype that seems to work fine and I'll include it in a separate comment.

**Comment by pekkaklarck** [comment]
Thanks for the PR and sorry for a late review. The related code is pretty complicated and it's thus impression that you got the feature done. As I commented in the review, there's one bug and the added new exceptions are pretty complicated. As I also mentioned, I created a simpler prototype that seems to work fine and this is how it looks like:
```diff
diff --git a/src/robot/errors.py b/src/robot/errors.py
index 4e1d3786a..2aaee2292 100644
--- a/src/robot/errors.py
+++ b/src/robot/errors.py
@@ -143,10 +143,12 @@ class ExecutionStatus(RobotError):
     def can_continue(self, context, templated=False):
         if context.dry_run:
             return True
-        if self.syntax or self.exit or self.skip or self.test_timeout:
+        if self.syntax or self.exit or self.test_timeout:
             return False
         if templated:
             return context.continue_on_failure(default=True)
+        if self.skip:
+            return False
         if self.keyword_timeout:
             return False
         return self.continue_on_failure or context.continue_on_failure()
diff --git a/src/robot/running/bodyrunner.py b/src/robot/running/bodyrunner.py
index bd17f2cba..db5f14616 100644
--- a/src/robot/running/bodyrunner.py
+++ b/src/robot/running/bodyrunner.py
@@ -56,9 +56,18 @@ class BodyRunner:
                 self._run = exception.can_continue(self._context, self._templated)
         if passed:
             raise passed
+        if errors and self._templated:
+            errors = self._handle_skip_with_templates(errors, result)
         if errors:
             raise ExecutionFailures(errors)
 
+    def _handle_skip_with_templates(self, errors, result):
+        if len(result.body) == 1 or not any(e.skip for e in errors):
+            return errors
+        if all(item.skipped for item in result.body):
+            raise ExecutionFailed('All iterations skipped.', skip=True)
+        return [e for e in errors if not e.skip]
+
```
A functional difference between this prototype and your implementation is that possible skips are not enumerated in test's message. As I wrote in the review, I don't think that's too important in general and without that functionality implementation gets simpler.

**Comment by pekkaklarck** [comment]
I think this looks so promising that I'll add #4426 to RF 7.2 scope. The change can be considered backwards incompatible, but I don't think it's so big change that it would require waiting for a major version.

File: atest/robot/running/skip.robot [hunk 1]
```diff
@@ -139,3 +139,15 @@ Suite setup and teardown are not run if all tests are unconditionally skipped or
     Check Test Case     Skip using robot:skip
     Check Test Case     Skip using --skip
     Length Should Be    ${suite.suites[0].tests}    2
+
+Skip Template One Skip And Pass
```
[the remark this answers is not in the document]
Comment (olesz) [reply atest/robot/running/skip.robot]: Separate test file has been created.

File: atest/testdata/running/skip/skip.robot [hunk 2]
```diff
@@ -270,8 +270,73 @@ Failing Test
...
+    [Template]    Skip template one skip and pass
+    1
+    2
+
+Skip Template Two Skips And Pass
+    [Documentation]    PASS Skips occurred:
+    ...
+    ...    1) 2 == 2
+    ...
+    ...    2) 4 == 4
```
[the remark this answers is not in the document]
Comment (olesz) [reply atest/testdata/running/skip/skip.robot:284]: This part was refactored/removed after using the provided prototype

File: src/robot/errors.py [hunk 3]
```diff
@@ -143,10 +143,10 @@ def continue_on_failure(self, continue_on_failure):
     def can_continue(self, context, templated=False):
         if context.dry_run:
             return True
-        if self.syntax or self.exit or self.skip or self.test_timeout:
-            return False
         if templated:
             return context.continue_on_failure(default=True)
+        if self.syntax or self.exit or self.skip or self.test_timeout:
+            return False
```
[the remark this answers is not in the document]
Comment (olesz) [reply src/robot/errors.py:152]: This part was refactored/removed after using the provided prototype

File: src/robot/errors.py [hunk 4]
```diff
@@ -230,6 +230,51 @@ def _get_attrs(self, errors):
...
+
+    def _get_attrs(self, errors):
+        return {
+            'test_timeout': any(e.test_timeout for e in errors),
+            'keyword_timeout': any(e.keyword_timeout for e in errors),
+            'syntax': any(e.syntax for e in errors),
+            'exit': any(e.exit for e in errors),
+            'continue_on_failure': all(e.continue_on_failure for e in errors),
+            'skip': all(e.skip for e in errors)
+        }
```
[the remark this answers is not in the document]
Comment (olesz) [reply src/robot/errors.py]: This part was refactored/removed after using the provided prototype

File: atest/testdata/running/skip/skip.robot [hunk 5]
```diff
@@ -270,8 +270,73 @@ Failing Test
...
+Skip Template Two Skips And Fail
+    [Documentation]    FAIL Skips occurred:
+    ...
+    ...    0) 2 == 2
+    ...
+    ...    1) 4 == 4
+    ...
+    ...    Also failure occurred:
+    ...
+    ...    fail
```
[the remark this answers is not in the document]
Comment (olesz) [reply atest/testdata/running/skip/skip.robot:301]: This part was refactored/removed after using the provided prototype

**Comment by olesz** [comment]
I did the requested modifications.

Please help on how to modify the documentation.

File: src/robot/errors.py [hunk 6]
```diff
@@ -144,9 +144,12 @@ def can_continue(self, context, templated=False):
         if context.dry_run:
             return True
         if self.syntax or self.exit or self.skip or self.test_timeout:
-            return False
+            if self.syntax or self.exit or self.test_timeout:
+                return False
```
Comment (pekkaklarck) [inline src/robot/errors.py:148]: These two lines seem to have wrong indentation level. I expect that some acceptance tests fail due to that.

File: src/robot/errors.py [hunk 7]
```diff
@@ -183,9 +186,9 @@ def _get(self, error, attr):
 
 class ExecutionFailures(ExecutionFailed):
 
-    def __init__(self, errors, message=None):
+    def __init__(self, errors, message=None, attrs=None):
         super().__init__(message or self._format_message(errors),
-                         **self._get_attrs(errors))
+                         **(attrs or self._get_attrs(errors)))
```
Comment (pekkaklarck) [inline src/robot/errors.py:191]: These changes shouldn't be needed anymore and should be reverted.

File: src/robot/errors.py [hunk 8]
```diff
@@ -230,7 +233,6 @@ def _get_attrs(self, errors):
     def get_errors(self):
         return self._errors
 
-
```
Comment (pekkaklarck) [inline src/robot/errors.py:233]: We use two empty lines between classes as PEP-8 recommends. Please add this line and the line a bit more down back.

File: atest/testdata/running/skip/skip_with_template.robot [hunk 9]
```diff
@@ -0,0 +1,27 @@
+*** Settings ***
+Library            skiplib.py
+
+*** Variables ***
+${TEST_OR_TASK}    Test
```
Comment (pekkaklarck) [inline atest/testdata/running/skip/skip_with_template.robot:5]: Neither this variable nor the library above don't seem to be used for anything and can be removed.

File: atest/testdata/running/skip/skip.robot [hunk 10]
```diff
@@ -270,6 +270,52 @@ Failing Test
...
+    [Template]    Skip template one skip and pass
+    1
+    2
+
+Skip Template Two Skips And Pass
+    [Documentation]    PASS Skips occurred:
+    ...
+    ...    1) 2 == 2
+    ...
+    ...    2) 4 == 4
```
Comment (pekkaklarck) [inline atest/testdata/running/skip/skip.robot:284]: All changes to this file should be reverted.

**Review by pekkaklarck — CHANGES_REQUESTED** [review]
Little cleanup still needed.

**Comment by pekkaklarck** [comment]
This looks really good, but there's some cleanup still needed for code and tests.

In addition to the above, the User Guide also needs to be updated. I looked where this should be explained, and decided to restructure the old content so that there's now a separate section there for templates with multiple iterations (see e9e75b7ce5f94da3530e14470d0cb4aa0358f510). That section already explains that execution continues after a failure with templates, and there could be a separate paragraph after that explaining the SKIP behavior. The needs to be also a note that the behavior has been changed in RF 7.2.

The User Guide is created using reStructuredText and you can locally generate it with our custom scrip by running `doc/userguide/ug2html.py create` to see how changes look like. If all this sounds intimidating, I can easily write the needed docs myself after this PR has been merged. We are anyway moving to Markdown and MkDocs in the near future, so learning the old tech stack doesn't give much value for you.

File: src/robot/errors.py (hunk 8 above)
Comment (olesz) [reply src/robot/errors.py]: OK, done.

File: src/robot/errors.py (hunk 7 above)
Comment (olesz) [reply src/robot/errors.py]: Indeed, forgot to revert this change.

File: src/robot/errors.py (hunk 6 above)
Comment (olesz) [reply src/robot/errors.py]: Ohh, copy-paste mistake. I was unlucky and the tests passed. Sorry for that.
Refactored.

**Comment by olesz** [comment]
I corrected all your findings.

Also created the modification to the user guide.

File: atest/robot/running/skip_with_template.robot [hunk 11]
```diff
@@ -0,0 +1,22 @@
+*** Settings ***
+Suite Setup     Run Tests  --skip skip-this --SkipOnFailure skip-on-failure    running/skip/
```
Comment (pekkaklarck) [inline atest/robot/running/skip_with_template.robot:2]: Both `--skip` and `--SkipOnFailure` are unnecessary as none of the executed tests have these tags. All options can be replaced with `${EMPTY}`.

File: doc/userguide/src/CreatingTestData/CreatingTestCases.rst [hunk 12]
```diff
@@ -948,6 +948,18 @@ are executed even if one or more of them fails. It is possible to use this
 kind of `continue on failure`_ mode with normal tests too, but with
 the templated tests the mode is on automatically.
 
+Also in case of templated tests all the rounds are executed even if one
+or more of them are skipped.
+The aggregated result of a templated test with skipped iterations is:
+
+* PASS if there were at lease one PASS iteration with any number of SKIPs
+* SKIP if all iterations were SKIP
+* FAIL if any of the iterations were FAIL
```
Comment (pekkaklarck) [inline doc/userguide/src/CreatingTestData/CreatingTestCases.rst:957]: Perhaps FAIL should be first.

**Review by pekkaklarck — APPROVED** [review]
Great work! There are some minor things to fix, but I'll do that myself after merging this PR.

**Pull request merged**